GSoC'25: Virtme-ng integration for CVMFS tests across kernels

ShivamMadlani opened the pull request:
This is a contribution done through GSoC, for more details on implementation and steps I followed, please refer to my blog post [here](https://shivammadlani.github.io/GSoC_project_blog/2025/10/03/extending-kernel-support.html)

This PR introduces a setup for automated testing on kernel builds for CVMFS test suite using Virtme-ng (a tool to build and boot to a kernel from its source code).

Key additions include:

- Shell scripts for building kernels and running tests (`build_kernel.sh`, `run.sh`, `guest/run_tests.sh`)
- Github workflow integration (`.github/workflows/ci_vng.yaml`)
- Multiple default kernel versions and configurations under `kernel/` and `configs/`
- Modular test cases in tests/ covering kernel checks, probe tests, fuse version checks, and more
- Logging and result tracking in `results/`
- Documentation

Benefits:

- Helps validate kernel functionality and CVMFS compatibility in a safe and isolated virtualized environment
- Early detection of integration issues between kernels and CVMFS ensures more stable releases
- Easy addition of kernel versions
- Centralized logging and result tracking

ShivamMadlani committed:
added initial tests and kernel img

ShivamMadlani committed:
added test runner for virtme-ng

ShivamMadlani committed:
added test failure logging system

ShivamMadlani committed:
extended to multiple kernels

ShivamMadlani committed:
added v5.12 kernel

ShivamMadlani committed:
added dummy test failing on v5.12

ShivamMadlani committed:
added virtio_blk support to kernel

ShivamMadlani committed:
added configs for cache mount point

ShivamMadlani committed:
- added cache disk and network for the guest
- refractored guest script

ShivamMadlani committed:
refractored guest script

ShivamMadlani committed:
added a probe test

ShivamMadlani committed:
forced non-root script execution in guests

ShivamMadlani committed:
updated mount for cache to default i.e. /var/lib/cvmfs

ShivamMadlani committed:
added a scipt for supporting custom kernel builds

ShivamMadlani committed:
exit on failure

ShivamMadlani committed:
added fuse protocol negotiation

ShivamMadlani committed:
added documentation

ShivamMadlani committed:
added virtme-ng testing CI

ShivamMadlani committed:
use .cache in virtme-ng CI

ShivamMadlani committed:
checks for .deb pachages

ShivamMadlani committed:
checks for mk-build-deps

ShivamMadlani committed:
make sure .cache exists

ShivamMadlani committed:
redirect stdout to logfile

ShivamMadlani committed:
added dynamic cpu check

ShivamMadlani committed:
capture args and forward them to new shell to be handled

ShivamMadlani committed:
added test case exclusion and test suite selection

ShivamMadlani committed:
dont hardcode paths

ShivamMadlani committed:
change test directory as passed by args, default: src/

ShivamMadlani committed:
update mount point name

ShivamMadlani committed:
added caching

ShivamMadlani committed:
simplify installation

ShivamMadlani committed:
utilize suite selection and exclusion

ShivamMadlani committed:
update test path

ShivamMadlani committed:
run from tests/ dir as some tests are path sensitive

ShivamMadlani committed:
enable kvm support on gh runners

ShivamMadlani committed:
add automount support to kernels

ShivamMadlani committed:
update readme

ShivamMadlani committed:
exit on failure

ShivamMadlani committed:
added usage

ShivamMadlani committed:
exit gracefully

ShivamMadlani committed:
remove redundant directory

ShivamMadlani committed:
dont hardcode paths

ShivamMadlani committed:
update readme

ShivamMadlani committed:
use cvmfs proxy

ShivamMadlani committed:
add squid server

ShivamMadlani committed:
only enable client tests

ShivamMadlani committed:
missing quote

ShivamMadlani committed:
try direct proxy

ShivamMadlani committed:
ignore failures for now

ShivamMadlani committed:
allow all tests to run instead of exitting immediately

ShivamMadlani committed:
use cvmfs proxy

ShivamMadlani committed:
use existing test_functions and explicity disable autofs on startup

vvolkl commented:
Looks very nice! Test failure of dockerized ci suite and macos  are unrelated (gitlab connectivity issues). I'll give it a closer look soon, as a first remark, let's not check in the binary kernel images in to the repository. I've added them to our download area, you can curl them in the ci script from there: 

https://ecsft.cern.ch/dist/cvmfs/caches/kernel/v5.12/bzImage
https://ecsft.cern.ch/dist/cvmfs/caches/kernel/v5.11/bzImage

ShivamMadlani committed:
removed kernel binaries

ShivamMadlani committed:
use wget to fetch kernel images

ShivamMadlani commented:
> Looks very nice! Test failure of dockerized ci suite and macos are unrelated (gitlab connectivity issues). I'll give it a closer look soon, as a first remark, let's not check in the binary kernel images in to the repository. I've added them to our download area, you can curl them in the ci script from there:
> 
> https://ecsft.cern.ch/dist/cvmfs/caches/kernel/v5.12/bzImage https://ecsft.cern.ch/dist/cvmfs/caches/kernel/v5.11/bzImage

Thanks! I have updated the setup to use the given url to fetch kernel Images, I chose wget over curl as it allows recursive downloads. This way if we plan to add more kernels in the future, adding them under https://ecsft.cern.ch/dist/cvmfs/caches/kernel/ with it's respective version subdirectory will be enough to enable it.

christge-sft reviewed (approved):
My feedback for this PR was given during the development. I think this PR is ready to be merged.

ShivamMadlani committed:
ensure apache server is running so proxy works

add dynamic selection of testcases and testsuites to be ran on CI

ShivamMadlani committed:
fix path

ShivamMadlani committed:
configure squid and clarify vm network routing

ShivamMadlani committed:
explicitly export TEST_PROXY variable

ShivamMadlani committed:
export CVMFS_TEST_PROXY from inside the VM instead of exporting beofre run.sh as vng vm doesnot inherit env vars

vvolkl commented:
A proxy really isn't required for these tests, it could also be removed and CVMFS_HTTP_PROXY set to DIRECT

ShivamMadlani committed:
use CVMFS_TEST_PROXY=DIRECT

ShivamMadlani committed:
report failure in CI if any one of the tests failed

christge-sft committed:
Remove kernel check

Kernel version 5.12 was hardcoded in the testing suit. This would
require adjusting the test to every different kernel tested which add
extra maintenance effort without great value as a test.

Signed-off-by: Georgios Christodoulis <christge@cern.ch>

christge-sft committed:
Remove squid support.

Since we are using a DIRECT proxy for our testing, squid was employed to
cache data for multiple kernel testing. Although, testing multiple
kernels (with or without squid) exceeds by far the acceptable duration,
hence we are reverting to a single kernel setup.

Signed-off-by: Georgios Christodoulis <christge@cern.ch>

christge-sft committed:
Specify a kernel version for testing

christge-sft committed:
Exclude non important tests for new kernel testing

vvolkl committed:
small fixes

vvolkl committed:
up

vvolkl committed:
use run.sh instead of guest/run_tests.sh

vvolkl committed:
fix docker mode

vvolkl committed:
add uname -a printout to vng script

vvolkl committed:
debug qemu

vvolkl committed:
fix initramfs issue

vvolkl committed:
up

vvolkl committed:
up

vvolkl committed:
fix kernel build

vvolkl committed:
fix kernel build

vvolkl committed:
up
